src/Raven.Server/Documents/DocumentDatabase.cs

```diff
@@ -618,25 +630,43 @@ private void ExecuteClusterTransactionOneByOne(List<ClusterTransactionCommand.Si
             try
             {
                 var index = command.Index;
+
                 var options = mergedCommands.Options[index];
                 if (exception == null)
                 {
-                    Task indexTask = null;
                     if (options.WaitForIndexesTimeout != null)
                     {
-                        indexTask = BatchHandler.WaitForIndexesAsync(DocumentsStorage.ContextPool, this, options.WaitForIndexesTimeout.Value,
+                        var indexTask = BatchHandler.WaitForIndexesAsync(DocumentsStorage.ContextPool, this, options.WaitForIndexesTimeout.Value,
                             options.SpecifiedIndexesQueryString, options.WaitForIndexThrow,
                             mergedCommands.LastDocumentEtag, mergedCommands.LastTombstoneEtag, mergedCommands.ModifiedCollections);
-                    }
 
-                    var result = new BatchHandler.ClusterTransactionCompletionResult
+                        var removeTask = ClusterTransactionWaiter.CreateTask(command.Options.TaskId, command.Index, out _);
+
+                        indexTask.ContinueWith(t =>
+                        {
+                            try
+                            {
+                                t.GetAwaiter().GetResult(); // Task t is already completed here
+                                ClusterTransactionWaiter.SetResult(options.TaskId, index, command.PreviousCount);
+                            }
+                            catch (Exception e)
+                            {
+                                ClusterTransactionWaiter.SetException(options.TaskId, index, e);
+                            }
+
+                            ThreadingHelper.InterlockedExchangeMax(ref LastCompletedClusterTransactionIndex, index);
+                            removeTask.Dispose();
+                        });
+                    }
+                    else
                     {
-                        Array = mergedCommands.Replies[index],
-                        IndexTask = indexTask,
-                    };
-                    ClusterTransactionWaiter.SetResult(options.TaskId, index, result);
+                        ClusterTransactionWaiter.SetResult(options.TaskId, index, command.PreviousCount);
+                        ThreadingHelper.InterlockedExchangeMax(ref LastCompletedClusterTransactionIndex, index);
```

Comment:
this should be out side this `if/else` I think

Reply:
Done